### DISCUSSION TITLE
Welcome to rosetta-dogecoin Discussions!

### DISCUSSION THREAD

**Discussion by patricklodder** [rosetta-dogecoin/rosetta-dogecoin#1] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by ithinkiborkedit** [comment]
How much refactoring is there involved considering how similar Doge is to Bitcoin(being a litecoin fork, which forked from Bitcoin).

I’m looking to get involved with the development community and Dogecoin seems like a cool project to be apart of. I’m a devops guy by trade and primarily have experience with python.

**Reply by patricklodder** [reply]
We actually reforked Dogecoin off Bitcoin, and then re-applied the Litecoin changes. So it's even better :-)

Not much refactoring is needed. Hopefully we can even remove some of the code that is now ported over and just use the upstream code - that would be ideal.

**Reply by ithinkiborkedit** [reply]
Brilliant!

**Reply by gooldbee** [reply]
> We actually reforked Dogecoin off Bitcoin, and then re-applied the Litecoin changes. So it's even better :-)
> 
> Not much refactoring is needed. Hopefully we can even remove some of the code that is now ported over and just use the upstream code - that would be ideal.

hello, I beg you, help me get my money back with dogechain info, save my life!

**Comment by jorellano** [comment]
Thanks Patrick, looking forward to this. Hopefully this will be the catalyst for more to help.

I'm a analyst/datascientist by trade, not sure if my skills and programming languages are useful at the moment, but i am happy to learn and help make dogecoin scale. Been a Shibe since 2014.

**Reply by michilumin** [reply]
In this case - watch what goes on here, just track the discussion and see if there's any insight you can offer. Contributions are not just code, so if you think it's of value, speak up.

**Comment by michilumin** [comment]
**Guys, we're going to try to limit all discussion of implementing this to here,**  Too much crosstalk in just sending Patrick private messages on Reddit, so we need to try to keep it orderly in this spot.  **Right now we're looking at what the requirements and needs are to integrate rosetta so please be patient with us. **  

Some assumptions (which we'll clarify as best we can here as we dig into this):

- Using Bitcoin as a model is probably valid since yes we inherit a lot from Bitcoin; so aside from things like fee rates, that's probably a good starting point.
- Even if you don't have direct coding skills I encourage you to take a look at what we're discussing and maybe research previous implementations;  it's always good to get a "30,000 foot perspective" vs being an engineer in the trenches. so, just be our reality check, "no stupid questions".

Really though we need to confine reddit PMs to absolute need because managing this is going to be rough without all eyes seeing requirements and progress. Stay tuned **here** for updates.

**Comment by michilumin** [comment]
Ok, so, it looks like the 'meat' of this that we want to focus on is

https://github.com/rosetta-dogecoin/rosetta-dogecoin/blob/main/bitcoin/client.go#L58-L85

Except for estimatesmartfee (which we're working on changing) - all of these functions should, for the most part, be a straight 1:1 with bitcoin (after changing the identifiers and denominations.) - so right now let's suggest folks take a look at that and their dogecoin equivalents:

getblock, getblockhash, blockchaininfo, pruneblockchain getrawmempool :
 https://github.com/dogecoin/dogecoin/blob/master/src/rpc/blockchain.cpp

getpeerinfo: 
https://github.com/dogecoin/dogecoin/blob/master/src/rpc/net.cpp

sendrawtransaction: 
https://github.com/dogecoin/dogecoin/blob/master/src/rpc/rawtransaction.cpp

estimatesmartfee:
(Several places, but likely will point to an override we are going to put in dogecoin.cpp soon.)

So if folks can start looking at this from both sides, that'd be great... (we are continuing to...)

**Reply by michilumin** [reply]
If anyone has any questions/curiosities though -- please post them here, lets start to discuss implementation.

**Reply by victorsk2019** [reply]
Hi.  This could sound too simplistic, but if you are looking for changing identifiers, denominations, names from code in rosetta-bitcoin for rosetta-dogecoin perhaps some of it could be sped up by using unix search and replace command:

    find . -type f -exec sed -i 's/this/that/g' {} \;  
 to automatically replace one specific pattern occurence in project to another?

**Reply by patricklodder** [reply]
Hi! Yes, this could work! But then, realize that in cases where we don't change any code other than a name, renaming it reduces maintainability long-term. So short term gain, long term loss. I'd always recommend the long term most sustainable solution.

**Comment by patricklodder** [comment]
I've added some issues describing work that need to be done and will continue adding as I run through an analysis of what is needed. See https://github.com/rosetta-dogecoin/rosetta-dogecoin/issues

**Comment by trevtravtrev** [comment]
Hello I'm a software engineer and my primary language is python. I want to pitch in on this project

**Reply by patricklodder** [reply]
I'd recommend you to look at issues and discussions here as people start interacting with the repository and just keep an eye out because definitely not everything needs hardcore golang programming - there's a lot of entry-level things that can be done, and even more generic integration work like docker, shell scripts, Makefiles - all that needs attention too. I have created a small guide on how to get started with helping this project under topic #7.

**Comment by xanimo** [comment]
Hi all,
Been doing dev work for about 20 years, looking forward to getting my hands dirty contributing to this with you all!  I've always been a lone wolf of sorts and only a few times worked within a public repo with others but I'm ready for the challenge and refresher!  Outside of that I'm a late adopter to $doge, only got into it last year but from what I've found and researched it is hands down my favorite within the crypto space 🚀
Cheers,
Dakoda

**Reply by michilumin** [reply]
Hey, look, imagine going on to Coinbase and doing a dogecoin transaction and knowing that your work is in play making that function.  Welcome, and if you have any questions on procedures in working on a public repo *please* don't be afraid to ask, there are no stupid questions.

**Reply by xanimo** [reply]
I am imagining it and it feels great!  Lol.  I will make sure to ask questions, thank you.  

I understand the general process I think, fork main repo, make a branch, pull from upstream and merge changes if need be, write code, add detailed description for commit, push to origin and then make pull req on main repo correct?

**Reply by michilumin** [reply]
That's correct -- if you want you can also commit and push to your repo and branch and then use github's cross-repo compare ('compare against another repository' i believe) and if it says 'ok to merge' then submit the PR that way, if you want a few more double checks along the way.  Either way also feel free to point folks at your repo to pull and build and give things a try, too.

**Reply by xanimo** [reply]
Cool thank you!  I just submitted my first PR, hopefully that was the solution to the issue.

**Comment by DennyDaVjncj** [comment]
What is the overall objective to developing this coin?

**Reply by michilumin** [reply]
Hi, this repo isn't for developing a new coin. It's for development of the Coinbase Rosetta API interface to allow Coinbase to readily talk to the (already-established) Dogecoin Core wallet/node.  

Or are you aware of that and asking "what's the point of continuing to work on dogecoin?" 

if it's the latter, this isn't really the venue for that; you can discuss that in /r/dogecoindev on reddit or in the dogecoin discord.

**Reply by DennyDaVjncj** [reply]
Thanks for the response @michilumin, it definitely isn't the latter. I have a huge interest in DogeCoin (owning a few hundred of em') & being a recent software engineering grad, it's certainly cool to discuss with those on the front line of development, no matter the aspect

**Comment by trevtravtrev** [comment]
Does anyone have any suggestions for getting up to speed on dogecoin's implementation to gain a working level understanding to be able to start implementing this API?

**Reply by xanimo** [reply]
'Google Fu'  I'd say just try and research and tinker with as much as you can.  Maybe try compiling it from source.  This project in particular is more so focused on the Rosetta API from Coinbase so maybe start there?

**Reply by michilumin** [reply]
I'd absolutely give a shot at compiling from source and seeing if you can get it running in your environment, first thing.  Then you can go on pretty quickly to interacting with dogecoind or just editing source files.

**Comment by zaridan** [comment]
Hi All!

I'm not a developer, but I am able to hack and copy/paste and work with repositories on things that are defined. I've contributed to Open Source projects going back to 2005 and would love to pitch in here in whatever ways I can.
Thanks to Ðogecoin Core Devs for getting this going, and for everyone old and new to the Community for jumping in. It will be super exciting to see Ðogecoin on Coinbase!

**Comment by coolcat2012** [comment]
I am a software developer who loves the Dogecoin and would hope to contribute the code to the repo here. Will learn to fork and set up the testing environment here locally and may need some help to speed up.

**Comment by victorsk2019** [comment]
Hi everyone,

I'm new to this particular area so I apologize in advance if it will sound like a total nonsense what I'm doing here (I'm also ~~new to GitHub~~ never used GitHub before so I'm really learning a lot!).

So I've set up rosetta-dogecoin repository on my GitHub account.  I then looked at making testing functionality work as there still seem to be references to bitcoin and not dogecoin.  So I've created separate 'dogecoin.ros' and modified 'config.json' to use 'dogecoin.ros' testnet.

As expected, I'm now getting errors when trying to run test, for example:

```
go test ./services/... ./indexer/... ./bitcoin/... ./dogecoin/...
--- FAIL: TestConstructionService (0.00s)
    construction_service_test.go:80: 
        	Error Trace:	construction_service_test.go:80
        	Error:      	Not equal: 
        	            	expected: &types.ConstructionDeriveResponse{AccountIdentifier:(*types.AccountIdentifier)(0xc000068140), Metadata:map[string]interface {}(nil)}
        	            	actual  : &types.ConstructionDeriveResponse{AccountIdentifier:(*types.AccountIdentifier)(0xc000068040), Metadata:map[string]interface {}(nil)}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -2,3 +2,3 @@
        	            	  AccountIdentifier: (*types.AccountIdentifier)({
        	            	-  Address: (string) (len=34) "DSpgzjPyfQB6ZzeSbMWpaZiTTxGf2oBCs4",
        	            	+  Address: (string) (len=44) "tdge1qj8qlgw4qtnxxgxl5nzs5n56fxkqg2e5rnqc6kn",
        	            	   SubAccount: (*types.SubAccountIdentifier)(<nil>),
        	Test:       	TestConstructionService 

```

I tried to modify 'construction_service_test.go' by replacing bitcoin public key "0325c9a4252789b31dbb3454ec647e9516e7c596bcde2bd5da71a60fab8644e438" with dogecoin's public key which I tried to generate from dogecoin's private keys defined in 'key_tests.cpp' from dogecoin-core  (I was using online bitcoin's public key generator tool and pasted in dogecoin's private key to produce public key).  But it didn't work, of course.  I think the error is because public key doesn't correspond to the dogecoin address I specified?  If so, how can the new public key for the dogecoin address be generated?

Thank you,
Victor.

**Reply by patricklodder** [reply]
So the problem with this one is that the address gets encoded with bech32 `tdge` prefix for testnet and your fixture is base58 `D` prefix for mainnet.

Currently bech32 is not fully supported with Dogecoin 1.14. If there are similar tests for base58 I would just disable these specific tests for now. If only bech32 is supported, we'll need to backport base58 into this repository - which will be fun :-)

**Reply by victorsk2019** [reply]
Hi.  Thank you for replying.  I tried to decode it as base58 earlier using online decoding tools but got back gibberish so that explains it :)

**Reply by victorsk2019** [reply]
Hi.  I replaced bitcoin addresses with (what I believe) should be Dogecoin's bech32-encoded address in construction_service_test.go (got these "tdge" addresses from information in error stack trace.  Anyways, the error message I get is:
```
Error:      	Expected nil, but got: &types.Error{Code:10, Message:"Unable to decode address", Description:(*string)(nil), Retriable:false, Details:map[string]interface {}{"context":"decoded address is of unknown format unable to decode address tdge1q3r8xjf0c2yazxnq9ey3wayelygfjxpfqmg6g08"}}
```

But after test process completes I get the following message:

   ```
     	Test:       	TestConstructionService
    construction_service_test.go:412: PASS:	SuggestedFeeRate(*context.emptyCtx,int64)
    construction_service_test.go:412: PASS:	SuggestedFeeRate(*context.emptyCtx,int64)
    construction_service_test.go:412: PASS:	SendRawTransaction(*context.emptyCtx,string)
    construction_service_test.go:413: PASS:	GetScriptPubKeys(*context.emptyCtx,[]*types.Coin)
    construction_service_test.go:413: PASS:	GetScriptPubKeys(*context.emptyCtx,[]*types.Coin)
FAIL
FAIL	github.com/rosetta-dogecoin/rosetta-dogecoin/services	1.083s
ok  	github.com/rosetta-dogecoin/rosetta-dogecoin/indexer	(cached)
ok  	github.com/rosetta-dogecoin/rosetta-dogecoin/bitcoin	(cached)
ok  	github.com/rosetta-dogecoin/rosetta-dogecoin/dogecoin	(cached)
FAIL
make: *** [Makefile:72: test] Error 1
```
I am not sure why "Unable to decode address" error comes up when the error messages themselves contain addresses with "tdge" prefix in them as suggested option?  Here's an example of one the messages included when running 'make test':

```
	  AccountIdentifier: (*types.AccountIdentifier)({
        	            	-  Address: (string) (len=34) "DSpgzjPyfQB6ZzeSbMWpaZiTTxGf2oBCs4",
        	            	+  Address: (string) (len=44) "tdge1qj8qlgw4qtnxxgxl5nzs5n56fxkqg2e5rnqc6kn",
```
I replaced "DSpgzjPyfQB6ZzeSbMWpaZiTTxGf2oBCs4" with "tdge1qj8qlgw4qtnxxgxl5nzs5n56fxkqg2e5rnqc6kn" but still getting "Unable to decode address" when using one with "tdge" prefix.  Any ideas why?

Thanks,
Victor.

**Reply by patricklodder** [reply]
I copied this into an issue ( #23 ), as it seems we need to look deeper into this. I think this is due to the services expecting bech32 rather than b58c addresses, not entirely sure yet. At least now it should get more visibility.

**Comment by Scerini** [comment]
Hello just looking to touch base and check status update. Given the current progress what can we expect as timeline?

**Reply by victorsk2019** [reply]
Hi.  Thank you for checking in.  I would love to help but find it an entirely new ecosystem to me or something I never worked with closely as a software developer.  I can relate some general concepts I worked with previously but mostly it's quite new to me.  It's a great learning experience but I think we may need some additional Doge core developers to assist us, if possible, so we can deploy Doge on Coinbase sooner.  Thanks :)

**Reply by godcloutier** [reply]
im here and golang is my specialty was busy trying to code a go wrapper around dogecoin dumb me ... lets finish this first then ill go back to my things after if I can help

**Comment by ghost** [comment]
Hey I would like to help, have some questions when I set up the dev environment:
1. I can run the docker which contains 1) the Dogecoin core c++ app running 2) the Golang rosetta code running. However, the core app always got stuck there after long time syncing blocks. is this because my Apple Mac laptop memory is too low? or it is due to the docker disk setting is too low which ended up disk full?
2. I tried to run the two apps without docker. I downloaded the Dogecoin core source code and try to compile and install on my Apple Mac laptop but always failed. Do I have to run the doge core in Linux instead of Mac? I saw the wiki page mentioned it is tested in AWS 16G env, does this mean I have to get a AWS EC2 host to do the development?
3. My goal right now is, run a doge standard node, and run the current rosetta code and do some testing with CLI. then see which functions are missing. I developed the ethereum things before using Golang, and used c++ for years.

**Reply by patricklodder** [reply]
1. With which parameters are you trying to sync? Mainnet? That's a lot of data (about 50Gb) and this is definitely problematic for testing (because if you'd want to test sync, you'd be downing that 50Gb again) Currently we're using testnet (which is 2Gb) and I'm about to request/perform integration of the regtest network so that we don't need to sync all those blocks during development but can keep it lean.

2. You don't need a huge node for running dogecoind, especially not for testnet, 2Gb ram and 2 cpu cores should do fine.  We use the binary released v1.14.3 dogecoind for this project (to not create cross-dependencies), so you'd want the same on your Mac (for now). You can self-compile this by cross-compiling from linux (or docker) with:

```bash
pushd depends
mkdir sdk-sources
mkdir SDKs
curl --location --fail https://bitcoincore.org/depends-sources/sdks/MacOSX10.11.sdk.tar.gz -o sdk-sources/MacOSX10.11.sdk.tar.gz
tar -C SDKs -xf sdk-sources/MacOSX10.11.sdk.tar.gz
make HOST=x86_64-apple-darwin11
popd
./autogen.sh
./configure --prefix=/build/depends/x86_64-apple-darwin11
make
```

**Reply by ghost** [reply]
Thank you Patrick, I am running the TestNet now and sync is slow and sometimes loses all the peers. I guess it will take days to download all the 2G blocks.
Looking forward to the  regtest...

**Reply by patricklodder** [reply]
Hmm that's not good. Both my testnet and mainnet nodes are stable nowadays, and I hadn't run into much trouble trunning the sync. I did hear of some problems on mainnet where people still get too many 1.14.2 nodes and sync is slow, but generally things should be predictable. I'm checking if maybe something with the testnet DNS seed is malfunctioning w/ the operator. Will take me a bit to find out if something is wrong, but that's

**Reply by ghost** [reply]
Sync is done and there is an error for the "getblock" API, as it passes the two paramaters 1) block hash 2) verbosity=2
set the verbosity to 2 will ask for the transactions in the block. as the current doge core API does not support verbosity=2, it failed. 
I temporarily remove the verbosity and only sent the block hash parameter, then I successfully got the block data.

the second issue is decoding the block data error, as the current rosetta code expects the transaction data in the block, where the we do not have. I fixed the issue by changing the rosetta code to NOT expect the transactions.

currently it is running and indexer is busy loading. I will do CLI testing after the indexer finish the job.

**Reply by patricklodder** [reply]
> the second issue is decoding the block data error, as the current rosetta code expects the transaction data in the block, where the we do not have. I fixed the issue by changing the rosetta code to NOT expect the transactions.

on `main` this should be fixed already through #19 though?

**Reply by ghost** [reply]
I see, can I get a small task to work on? is there a todo list in the backlog?

**Reply by gooldbee** [reply]
> Hmm that's not good. Both my testnet and mainnet nodes are stable nowadays, and I hadn't run into much trouble trunning the sync. I did hear of some problems on mainnet where people still get too many 1.14.2 nodes and sync is slow, but generally things should be predictable. I'm checking if maybe something with the testnet DNS seed is malfunctioning w/ the operator. Will take me a bit to find out if something is wrong, but that's

hello, I beg you, help me get my money back with dogechain info, save my life!

**Comment by gooldbee** [comment]
@patricklodder  https://github.com/rosetta-dogecoin/rosetta-dogecoin/issues/55
